homeassistant/helpers/event.py

```diff
@@ -322,37 +324,65 @@ def remove_listener():
 
 @callback
 @bind_hass
-def async_track_utc_time_change(hass, action, year=None, month=None, day=None,
+def async_track_utc_time_change(hass, action,
                                 hour=None, minute=None, second=None,
                                 local=False):
     """Add a listener that will fire if time matches a pattern."""
     # We do not have to wrap the function with time pattern matching logic
     # if no pattern given
-    if all(val is None for val in (year, month, day, hour, minute, second)):
+    if all(val is None for val in (hour, minute, second)):
         @callback
         def time_change_listener(event):
             """Fire every time event that comes in."""
             hass.async_run_job(action, event.data[ATTR_NOW])
 
         return hass.bus.async_listen(EVENT_TIME_CHANGED, time_change_listener)
 
-    pmp = _process_time_match
-    year, month, day = pmp(year), pmp(month), pmp(day)
-    hour, minute, second = pmp(hour), pmp(minute), pmp(second)
+    matching_seconds = dt_util.parse_time_expression(second, 0, 59)
+    matching_minutes = dt_util.parse_time_expression(minute, 0, 59)
+    matching_hours = dt_util.parse_time_expression(hour, 0, 23)
+
+    next_time = None
+
+    def calculate_next(now):
+        """Calculate and set the next time the trigger should fire."""
+        nonlocal next_time
+
+        localized_now = dt_util.as_local(now) if local else now
+        next_time = dt_util.find_next_time_expression_time(
+            localized_now, matching_seconds, matching_minutes,
+            matching_hours)
+
+    # Make sure rolling back the clock doesn't prevent the timer from
+    # triggering
+    last_now = None
 
     @callback
     def pattern_time_change_listener(event):
         """Listen for matching time_changed events."""
+        nonlocal next_time, last_now
+
         now = event.data[ATTR_NOW]
 
-        if local:
-            now = dt_util.as_local(now)
+        # Prevent "can't compare offset-naive and offset-aware" errors,
+        # let's make everything timezone-aware
+        if now.tzinfo is None:
+            # pylint: disable=no-value-for-parameter
+            now = pytz.UTC.localize(now)
+
+        if last_now is None or now < last_now:
+            # Time rolled back or next time not yet calculated
+            calculate_next(now)
 
-        # pylint: disable=too-many-boolean-expressions
-        if second(now.second) and minute(now.minute) and hour(now.hour) and \
-           day(now.day) and month(now.month) and year(now.year):
+        last_now = now
 
-            hass.async_run_job(action, now)
+        if next_time <= now:
+            hass.async_run_job(action, event.data[ATTR_NOW])
+            base = now
+            if next_time == now:
+                # Make sure not to trigger on duplicate times
+                base = now + timedelta(seconds=1)
+            calculate_next(base)
 
     return hass.bus.async_listen(EVENT_TIME_CHANGED,
```

Comment:
Wouldn't it make more sense to just call `track_utc_point_in_time` now that we have a point in time ? Then if we optimize that function, we optimize it all ?

Reply:
I agree this should be refactored into one system.

There are two problems with doing it right now:

 * First is the system time rolling backward. Previously, that would have been handled correctly (because each second was checked), but `track_utc_point_in_time` would break it.
 * Also, the starting `next_time` is calculated lazily by the value obtained from the first `time_changed` event. Using `track_utc_point_in_time` would mean that it would need to be calculated using `dt_util.utcnow()`, which would complicate the tests a little.

Fundamentally, the problem is that there are several kinds of "run this at that time":

 * Run this in exactly 10 minutes, no matter system time changes.
 * Run this in exact 10 minute intervals, no matter system time changes.
 * Run this at exactly 14:30 of october the 8th.
 * Run this the next time at 14:30, or if the system time does some weird stuff, re-schedule it to this other time.

And of course there are a few system time scenarios that complicate things a whole bunch:

 * System time abruptly being rolled backwards. If for example the NTP server changes (my windows laptop had this after each reboot, lots of software just crashed)
 * System time abruptly being in the future. The scenario above or the system going to sleep for a while.

Each of the scenarios needs its own kind of scheduling behavior. I plan on writing up a quick PR showing what my proposed scheduler would look like.

Reply:
Well the question is what is the impact for our end users. I think that we should indeed handle it, however it is adding a lot of complexity. Complexity that we should at least describe better in the comments why it's there.